### ISSUE TITLE
Java: CWE-600 Uncaught servlet exception

### ISSUE THREAD

**Issue by luchua-bc** [github/securitylab#143]
## CVE ID(s)

*List the CVE ID(s) associated with this vulnerability. GitHub will automatically link CVE IDs to the [GitHub Advisory Database](https://github.com/advisories).*

## Report

*Describe the vulnerability. Provide any information you think will help GitHub assess the impact your query has on the open source community.*

Even though method signatures of a servlet include throws `IOException`, `ServletException`, and runtime exceptions don't need to be caught for the program to be compiled, it's a bad idea to let such exceptions be thrown. 

Failure to catch exceptions in a servlet could leave a system in a vulnerable state, possibly resulting in denial-of-service attacks, or the exposure of sensitive information because when a servlet throws an exception, the servlet container typically sends debugging information back to the user. And that information could be very valuable to an attacker.

The CodeQL query in the related PR detects method calls within servlet methods that can throw IOException and RuntimeException but those exceptions are not caught then processed.

I've tested the query against some popular open-source software including Apache ActiveMQ.

Relevant PR [#3839](https://github.com/github/codeql/pull/3839)

- [x] Are you planning to discuss this vulnerability submission publicly? (Blog Post, social networks, etc). *We would love to have you spread the word about the good work you are doing*

## Result(s)

*Provide at least one useful result found by your query, on some revision of a real project.*

Apache ActiveMQ is a popular software. Running the query against its latest version got the following results:
<img width="1039" alt="activemq_lgtm" src="https://user-images.githubusercontent.com/37377780/86053035-e6fb3300-ba25-11ea-847e-1f46e48d5ec1.png">

To double confirm query results, I did a POC by installing the latest version 5.15.13 on my VM then exploiting vulnerable services. Sending an invalid number as a request parameter to the vulnerable service triggered the uncaught `NumberFormatException` runtime exception as expected:
<img width="959" alt="activemq_web_parseint" src="https://user-images.githubusercontent.com/37377780/86053183-2033a300-ba26-11ea-92f2-b35f8ac186ec.png">

Similarly passing an invalid UTF-8 character as a request parameter to the vulnerable service triggered the uncaught encoding `IOException` as expected:
<img width="1000" alt="activemq_web_encode" src="https://user-images.githubusercontent.com/37377780/86053405-730d5a80-ba26-11ea-8915-befef9059797.png">

In summary, the tests confirmed results found by my query.

**Comment by ghsecuritylab** [comment]
Your submission is now in status **SecLab review**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by pwntester** [comment]
Hi @luchua-bc, thanks for the submission!

We think the query could be improved by checking the `web.xml` files and making sure there is no `<error-page>` defined in order to prevent FPs.
In addition, we spotted some issues where the exception being thrown is actually handled in a catch block, eg: https://lgtm.com/projects/g/OWASP/Benchmark/snapshot/6479aeffa0531427b9b227721b3b342236f35296/files/src/main/java/org/owasp/benchmark/testcode/BenchmarkTest00001.java?sort=name&dir=ASC&mode=heatmap

Could you please take a look?

**Comment by luchua-bc** [comment]
Hi @pwntester,

Thanks for reviewing the PR. 

For the issue with the OWASP class, I think it's caused by a missing inheritance relationship in <code>javax.servlet.http.HttpServlet</code>. Currently it doesn't inherit any class, which has:
```
public abstract class HttpServlet {
```

However, it should be:
```
public abstract class HttpServlet extends GenericServlet {
```

Then <code>GenericServlet</code> implements the <code>Servlet</code> interface. Without this inheritance relationship, the following check in the query will fail:
```
isServletMethod(ma.getEnclosingCallable())
```

That is, the current stubs of servlet API in the repository don't maintain the relationship between <code>HttpServlet</code> and <code>Servlet</code>.

I will add the missing code, which will help this PR and other JEE PRs, together with the <code>web.xml</code> check for this PR.

**Comment by luchua-bc** [comment]
Actually the code I committed already has <code>GenericServlet</code> that maintains the inheritance relationship, which is missing in main since my PR hasn't been merged yet. It will explain why if you ran this query with the servlet API stubs in main.

**Comment by pwntester** [comment]
Thanks, that was quick!

**Comment by ghsecuritylab** [comment]
Your submission is now in status **CodeQL review**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **SecLab review**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **SecLab finalize**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **CodeQL review**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **SecLab finalize**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **CodeQL review**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **SecLab finalize**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Pay**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by xcorail** [comment]
Created Hackerone report 1052793 for bounty 263280 : [143] Java: CWE-600 Uncaught servlet exception

**Comment by ghsecuritylab** [comment]
Your submission is now in status **Closed**.

_For information, the evaluation workflow is the following:
CodeQL initial assessment > SecLab review > CodeQL review > SecLab finalize > Pay > Closed_

**Comment by luchua-bc** [comment]
Thanks @xcorail for the quick turn-around and the bounty.
